web/ui/src/common/baseFactory.js

```diff
@@ -0,0 +1,107 @@
+// baseFactory
+
+(function() {
+    'use strict';
+
+    var UPDATE_FREQUENCY = 3000;
+
+    var $q, $interval;
+
+    angular.module('baseFactory', []).
+    factory("baseFactory", ["$rootScope", "$q", "resourcesFactory", "$interval",
+    function($rootScope, _$q, resourcesFactory, _$interval){
+
+        $q = _$q;
+        $interval = _$interval;
+
+        return BaseFactory;
+    }]);
+
+    // TODO make update frequency configurable
+    // TODO - default ObjConstructor
+    function BaseFactory(ObjConstructor, updateFn){
+        this.objMap = {};
+        this.objArr = [];
+        this.updateFn = updateFn;
+        this.ObjConstructor = ObjConstructor;
+    }
+
+    BaseFactory.prototype = {
+        constructor: BaseFactory,
+
+        update: function(){
+            var deferred = $q.defer();
+            this.updateFn()
+                .success((data, status) => {
+                    var included = [];
+
+                    for(let id in data){
+                        let obj = data[id];
+
+                        // update
+                        if(this.objMap[obj.ID]){
+                            this.objMap[obj.ID].update(obj);
+
+                        // new
+                        } else {
+                            this.objMap[obj.ID] = new this.ObjConstructor(obj);
+                            this.objArr.push(this.objMap[obj.ID]);
+                        }
+
+                        included.push(obj.ID);
+                    }
+
+                    // delete
+                    if(included.length !== Object.keys(this.objMap).length){
+                        // iterate objMap and find keys
+                        // not present in included list
+                        for(let id in this.objMap){
+                            if(included.indexOf(id) === -1){
+                                this.objArr.splice(this.objArr.indexOf(this.objMap[id], 1));
+                                delete this.objMap[id];
+                            }
+                        }
+                    }
+
+                    deferred.resolve();
+                });
+            return deferred.promise;
+        },
+
+        activate: function(){
```

Comment:
Can this be simplified to something like this just to avoid stopping/starting a new timer when the promise is basically the same each time?
```
activate: function() {
    if (!this.updatePromise) {
        this.updatePromise = $interval(this.update.bind(this), UPDATE_FREQUENCY);
   }
}
```